File: apps/api/src/callcraft_engine/ssrf.py
import ipaddress
import socket
from typing import Union
from urllib.parse import urlparse


class SsrfError(Exception):
    pass


def validate_url_ip(url_str: str) -> str:
    """Validates target URL against SSRF rules before making HTTP request."""
    try:
        parsed = urlparse(url_str)
    except Exception:
        raise SsrfError("Invalid host string")

    if parsed.scheme not in ("http", "https"):
        raise SsrfError("Invalid URL scheme: only HTTP and HTTPS are permitted")

    host = parsed.hostname
    if not host:
        raise SsrfError("Invalid host string")

    port = parsed.port or (443 if parsed.scheme == "https" else 80)

    try:
        addr_info = socket.getaddrinfo(host, port)
    except socket.gaierror:
        raise SsrfError(f"DNS resolution failed for host: {host}")

    for family, socktype, proto, canonname, sockaddr in addr_info:
        ip_str = sockaddr[0]
        try:
            ip = ipaddress.ip_address(ip_str)
        except ValueError:
            continue

        if is_private_or_restricted_ip(ip):
            raise SsrfError(f"SSRF Protection: Access to private/loopback/cloud IP address {ip} is forbidden")

    return url_str


def is_private_or_restricted_ip(ip: Union[ipaddress.IPv4Address, ipaddress.IPv6Address]) -> bool:
    if ip.is_loopback or ip.is_private or ip.is_link_local or ip.is_unspecified:
        return True
    if isinstance(ip, ipaddress.IPv4Address):
        if ip.is_multicast:
            return True
    return False

File: crates/ocr-engine/src/ssrf.rs
use std::net::{IpAddr, ToSocketAddrs};
use thiserror::Error;
use url::Url;

#[derive(Error, Debug)]
pub enum SsrfError {
    #[error("Invalid URL scheme: only HTTP and HTTPS are permitted")]
    InvalidScheme,
    #[error("Invalid host string")]
    InvalidHost,
    #[error("DNS resolution failed for host: {0}")]
    DnsResolutionFailed(String),
    #[error("SSRF Protection: Access to private/loopback/cloud IP address {0} is forbidden")]
    ForbiddenIpAddress(IpAddr),
}

/// Validates target URL against SSRF rules before making HTTP request
pub fn validate_url_ip(url_str: &str) -> Result<Url, SsrfError> {
    let url = Url::parse(url_str).map_err(|_| SsrfError::InvalidHost)?;

    // 1. Check scheme
    match url.scheme() {
        "http" | "https" => {}
        _ => return Err(SsrfError::InvalidScheme),
    }

    // 2. Resolve DNS IP Addresses
    let host = url.host_str().ok_or(SsrfError::InvalidHost)?;
    let port = url.port_or_known_default().unwrap_or(80);
    let socket_str = format!("{}:{}", host, port);

    let addrs = socket_str
        .to_socket_addrs()
        .map_err(|_| SsrfError::DnsResolutionFailed(host.to_string()))?;

    // 3. Verify every resolved IP against private / loopback / cloud metadata ranges
    for addr in addrs {
        let ip = addr.ip();
        if is_private_or_restricted_ip(ip) {
            return Err(SsrfError::ForbiddenIpAddress(ip));
        }
    }

    Ok(url)
}

fn is_private_or_restricted_ip(ip: IpAddr) -> bool {
    match ip {
        IpAddr::V4(ipv4) => {
            ipv4.is_loopback()                           // 127.0.0.0/8
                || ipv4.is_private()                     // 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
                || ipv4.is_link_local()                  // 169.254.0.0/16 (Includes AWS Metadata 169.254.169.254)
                || ipv4.is_broadcast()                   // 255.255.255.255
                || ipv4.is_documentation()               // 192.0.2.0/24, etc.
                || ipv4.is_unspecified()                 // 0.0.0.0
        }
        IpAddr::V6(ipv6) => {
            ipv6.is_loopback()                           // ::1
                || ipv6.is_unspecified()                 // ::
        }
    }
}
